mate-panel/panel-run-dialog.c

```diff
@@ -192,13 +195,23 @@ _panel_run_save_recent_programs_list (PanelRunDialog   *dialog,
 			guint pos;
 			for (pos = 0; pos < items_added / 2; pos++)
 			{
-				gchar const * tmp = items[pos + 1];
+				gchar *tmp = items[pos + 1];
 				items[pos + 1] = items[items_added - pos];
 				items[items_added - pos] = tmp;
 			}
 		}
-		items[history_max_size < items_added + 1 ? history_max_size : items_added + 1] = NULL;
+
+		if (history_max_size < items_added + 1) {
+			g_free (items[history_max_size]);
+			items[history_max_size] = NULL;
+		} else {
+			items[items_added + 1] = NULL;
+		}
 		g_settings_set_strv (dialog->settings, PANEL_RUN_HISTORY_KEY, items);
+
+		/* free the items */
+		for (; items_added; items_added--)
+			g_free (items[items_added]);
```

Comment:
[g_strfreev](https://developer.gnome.org/glib/stable/glib-String-Utility-Functions.html#g-strfreev)

Reply:
No, because you need not free the array itself, only its elements.

Reply:
you can also create an array dynamically and copy the command as well, ie do not assume that the last command is the current one (items[0]), doing it simplifies the code ;)

Reply:
```diff
diff --git a/mate-panel/mate-submodules b/mate-panel/mate-submodules
index dd9bd34..580a612 160000
--- a/mate-panel/mate-submodules
+++ b/mate-panel/mate-submodules
@@ -1 +1 @@
-Subproject commit dd9bd3436f2cb6410b04eefbd69306d93cd07b30
+Subproject commit 580a612e236053bddf552cc6819e49c508de4aab
diff --git a/mate-panel/panel-run-dialog.c b/mate-panel/panel-run-dialog.c
index f6fa95b..59c946a 100644
--- a/mate-panel/panel-run-dialog.c
+++ b/mate-panel/panel-run-dialog.c
@@ -172,20 +172,23 @@ _panel_run_save_recent_programs_list (PanelRunDialog   *dialog,
                model = gtk_combo_box_get_model (GTK_COMBO_BOX (entry));
 
                /* reasonable upper bound for zero-terminated array with new command */
-               gchar const *items[gtk_tree_model_iter_n_children (model, NULL) + 2];
-               guint        items_added = 0;
+               gchar **items;
+               guint   items_added = 0;
 
-               items[0] = lastcommand;
+               items = g_new0 (gchar*, gtk_tree_model_iter_n_children (model, NULL) + 2);
+               items[0] = g_strdup (lastcommand);
 
                if (gtk_tree_model_get_iter_first (model, &iter)) {
                        char *command;
                        do {
                                gtk_tree_model_get (model, &iter, 0, &command, -1);
-                               if (g_strcmp0 (command, lastcommand) == 0)
+                               if (g_strcmp0 (command, lastcommand) == 0) {
+                                       g_free (command);
                                        continue;
+                               }
                                items[items_added + 1] = command;
                                items_added++;
-                       } while (gtk_tree_model_iter_next (model, &iter));
+                       } while ((gtk_tree_model_iter_next (model, &iter)) && (items_added < history_max_size-1));
                }
 
                if (history_reverse) {
@@ -197,8 +200,9 @@ _panel_run_save_recent_programs_list (PanelRunDialog   *dialog,
                                items[items_added - pos] = tmp;
                        }
                }
-               items[history_max_size < items_added + 1 ? history_max_size : items_added + 1] = NULL;
+
                g_settings_set_strv (dialog->settings, PANEL_RUN_HISTORY_KEY, items);
+               g_strfreev (items);
        }
 }
 
```